alternator/executor.cc

```diff
@@ -5860,32 +5871,43 @@ static lw_shared_ptr<keyspace_metadata> create_keyspace_metadata(std::string_vie
     // word "none") will ask to disable tablets.
-    // If we make this tag a permanent feature, it will get a "system:" prefix -
-    // until then we give it the "experimental:" prefix to not commit to it.
-    static constexpr auto INITIAL_TABLETS_TAG_KEY = "experimental:initial_tablets";
-    // initial_tablets currently defaults to unset, so tablets will not be
-    // used by default on new Alternator tables. Change this initialization
-    // to 0 enable tablets by default, with automatic number of tablets.
+    // When vnodes are asked for by the tag value, but tablets are enforced by config,
+    // throw an exception to the client.
     std::optional<unsigned> initial_tablets;
     if (feat.tablets) {
         auto it = tags_map.find(INITIAL_TABLETS_TAG_KEY);
         if (it != tags_map.end()) {
-            // Tag set. If it's a valid number, use it. If not - e.g., it's
-            // empty or a word like "none", disable tablets by setting
-            // initial_tablets to a disengaged optional.
+            // The per-table tag is set.
             try {
-                initial_tablets = std::stol(tags_map.at(INITIAL_TABLETS_TAG_KEY));
+                // If the tag is a valid number, use it.
+                initial_tablets = std::stol(it->second);
+                elogger.info("Following {} tag, Alternator will attempt to create a keyspace {} with tablets, initially {} of them (0 means default number).", INITIAL_TABLETS_TAG_KEY, keyspace_name, initial_tablets.value());
             } catch(...) {
+                // If the tag is not a number - e.g., it's empty or is a word like "none",
+                // disable tablets and use vnodes by setting initial_tablets to a disengaged optional.
+                // When vnodes are asked for but tablets are enforced by config, throw.
+                if(tablets_mode == db::tablets_mode_t::mode::enforced) {
+                    throw api_error::validation(fmt::format("Creating keyspace {} with vnodes forbidden - tablets are enforced in configuration.", keyspace_name));
+                }
                 initial_tablets = std::nullopt;
+                elogger.info("Following {} tag containing non-numerical value, Alternator will attempt to create a keyspace {} with vnodes.", INITIAL_TABLETS_TAG_KEY, keyspace_name);
+            }
+        } else {
+            // No per-table tag present, use the value from config
+            if(tablets_mode == db::tablets_mode_t::mode::enabled || tablets_mode == db::tablets_mode_t::mode::enforced) {
+                initial_tablets = 0;
+                elogger.info("Following the `tablets_mode_for_new_keyspaces` flag from the settings, Alternator will attempt to create a keyspace {} with tablets.", keyspace_name);
+            } else {
+                elogger.info("Following the `tablets_mode_for_new_keyspaces` flag from the settings, Alternator will attempt to create a keyspace {} with vnodes.", keyspace_name);
```

Comment:
I see you didn't put `initial_tablets = std::nullopt;`. I agree that it's not needed, because that's the default value of this variable, but then we also don't need to do it above, right?